mage/packages.go

```diff
@@ -0,0 +1,982 @@
+		exclude:    []string{"**/*.tf"},
+		srcDir:     "assets/kubernetes",
+		deps: []string{
+			pkgGravity.Locator(),
+			pkgTeleport.Locator(),
+			pkgFio.Locator(),
+			pkgPlanet.Locator(),
+			pkgWebAssets.Locator(),
+			pkgRBAC.Locator(),
+			pkgDNSApp.Locator(),
+			pkgLoggingApp.Locator(),
+			pkgMonitoringApp.Locator(),
+			pkgIngressApp.Locator(),
+			pkgStorageApp.Locator(),
+			pkgBandwagonApp.Locator(),
+			pkgTillerApp.Locator(),
+			pkgSiteApp.Locator(),
+		},
+		force: true,
+	}
+)
+
+func (Package) Telekube() (err error) {
+	mg.Deps(Build.Go, Package.K8s)
+
+	m := root.Target("package:telekube")
+	defer func() { m.Complete(err) }()
+
+	return trace.Wrap(pkgTelekube.localAppImport(m, consistentStateDir()))
+}
+
+func (Package) K8s() (err error) {
+	mg.Deps(Build.Go, Package.GravityPackage, Package.Teleport, Package.Fio, Package.Planet, Package.Web,
+		Package.Site, Package.Monitoring, Package.Logging, Package.Ingress, Package.Storage, Package.Tiller,
+		Package.Rbac, Package.DNS, Package.Bandwagon)
+
+	m := root.Target("package:k8s")
+	defer func() { m.Complete(err) }()
+
+	return trace.Wrap(pkgKubernetes.localAppImport(m, consistentStateDir()))
+}
+
+func (Package) GravityPackage() (err error) {
+	mg.Deps(Build.Go)
+
+	m := root.Target("package:gravity-package")
+	defer func() { m.Complete(err) }()
+
+	// the gravity package operates directly on the version state directory, so use the shared lock
+	sharedStateMutex.Lock()
+	defer sharedStateMutex.Unlock()
+
+	gravityPackage := fmt.Sprint("gravitational.io/gravity:", buildVersion)
+
+	_, err = m.Exec().Run(context.TODO(),
+		consistentGravityBin(),
+		"--state-dir", consistentStateDir(),
+		"package",
+		"delete",
+		gravityPackage,
+		"--force",
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	_, err = m.Exec().Run(context.TODO(),
+		consistentGravityBin(),
+		"--state-dir", consistentStateDir(),
+		"package",
+		"import",
+		consistentGravityBin(),
+		gravityPackage,
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (Package) Teleport() (err error) {
+	mg.Deps(Build.Go)
+
+	m := root.Target("package:teleport")
+	defer func() { m.Complete(err) }()
+
+	cachePath := filepath.Join("build/apps", fmt.Sprint("teleport.", pkgTeleport.version, ".tar.gz"))
+
+	_, err = os.Stat(cachePath)
+	if !os.IsNotExist(err) {
+		m.SetCached(true)
+		return trace.Wrap(pkgTeleport.ImportPackage(m, cachePath))
+	}
+
+	tmpDir, err := ioutil.TempDir("", "build-teleport")
+	if err != nil {
+		return trace.ConvertSystemError(err)
+	}
+
+	defer os.RemoveAll(tmpDir)
+
+	m.Println("  tmpDir:", tmpDir)
+	m.Println()
+
+	buildDir := filepath.Join(tmpDir, "build")
+
+	for _, dir := range []string{
+		filepath.Join(buildDir, "/rootfs/usr/bin"),
+		filepath.Join(buildDir, "/rootfs/usr/share/teleport"),
+		tmpDir,
+	} {
+		err = os.MkdirAll(dir, 0755)
+		if err != nil {
+			return trace.ConvertSystemError(err)
+		}
+	}
+
+	_, err = m.Exec().SetWD(tmpDir).Run(context.TODO(),
```

Comment:
Looks like this always builds Teleport from source? In the current Makefile we first try to download a pre-built asset from S3. Same for monitoring, logging and other apps. Or is that logic elsewhere?

Reply:
IIRC for stuff that's in golang, the idea is to have a working GOCACHE. So building teleport should be relatively quick off the gocache versus the added complexity of trying to download a built binary from somewhere else.

So atleast as it stands today, magnet depends more on local caching, and doesn't do the remote caching stuff. Means a slower build every once in awhile, but on average it's much faster without the try to download if possible logic.